Add support for Account configuration (PBID-727, #1395)

laurb9 opened the pull request:
- Defines the Account object with per-publisher settings
- Adds a cfg.AccountDefaults for global settings that are overrideable per publisher account
- Can use cfg.AccountDefaults.Disabled=true to require a valid host-defined account not just any non-empty value
- Makes the resolved account object available to auction and analytics
- (optional) Cache TTL can now be controlled at account level
- Uses fetcher framework to retrieve accounts from individual <id>.json files or HTTP API
  (database to be added later)
- JSON account files are merged with AccountDefaults so they may specify overrides only for easier migration

Example configuration for accounts stored in `stored_requests/data/by_id/accounts/<id>.json`
```yaml
 accounts:
   filesystem:
     enabled: true
     directorypath: "./stored_requests/data/by_id"
```

See this unit test for how the flags control the various levels of validation supported:
https://github.com/prebid/prebid-server/blob/16f9f07b7eba94004748603282537e683a566bc2/endpoints/openrtb2/auction_test.go#L1841-L1872

hhhjort commented on endpoints/openrtb2/auction.go:
```diff
@@ -1282,14 +1285,25 @@ func effectivePubID(pub *openrtb.Publisher) string {
+	if account.Disabled {
+		if account.ID == config.UnknownAccount.ID {
+			err = error(&errortypes.AcctRequired{
+				Message: fmt.Sprintf("Prebid-server has been configured to discard requests that don't come with an Account ID. Please reach out to the prebid server host."),
```
The "account" ID is pulled from the openRTB "publisher" ID. Probably best to word the messages appropriately to reduce confusion, as there is no explicit "account" field in openRTB.

laurb9 replied:
I agree wrt confusion. With PBS' concept of "account" overlapping that of "publisher", where to reference one or the other is a bit of gray area. I've preserved the previous messaging in this case, but I can change it.

bsardo commented:
Hi @laurb9,
Sorry for the delay. I had been meaning to sync up with you so we can collaborate on the account service feature.
What you have looks great and is an approach that will work. However, as you noted, the caching piece is stored request specific and a refactor would really help. I’m wondering if we should take this opportunity to create a generic fetcher package that can initially be used just for accounts but can eventually be used for stored requests and whatever other data sources come about in the future. We could carry over some of the design approaches in the current fetcher logic to this new package. What are your thoughts on that?
If you’re interested in going down this path, that’s great! If not, that's ok too; you could continue going down the path you’re on and we can look into a more generic fetcher afterwards.

I'll open a draft PR shortly to capture what a generic fetcher might look like.

Also, I’d like you to be aware of the ticker task I created in https://github.com/prebid/prebid-server/pull/1394. I will be opening another PR shortly that addresses some initial/delta fetch bugs and I may end up refactoring the db and http fetchers a bit to use the ticker task for the periodic piece to simplify the logic and improve testability.

laurb9 commented:
Hi @bsardo, thanks for taking a look.

I like the idea of a generic fetcher. Ultimately what they do is KV retrieval by lists of keys, are type agnostic, and requests, imps, accounts and category maps do not actually need specific functions. 

Then there's the issue of combining stored requests and imps. Having them in a single http or sql query is probably intended to reduce overall latency, but I wonder if proper application of cache can't alleviate that. Retrieving them separately might improve code readability in other places.

So if a generic fetcher might work, I'd like to help if my timeline allows. Otherwise, I want to make sure the changes I am proposing have an easy migration path to that.

Regarding #1394, is your intention to replace the cache push event model with periodic refreshing of the data ?

bsardo commented:
Hi @laurb9,

I would still use the cache push event model. I just wanted to refactor the polling event producers (e.g. `stored_requests/events/Postgres/polling.go`) to use the ticker task.

Regarding the generic fetcher, I don't know if it makes sense for the entire fetcher to be generic. I think a generic storage layer and generic cache components would work. We could then create a fetcher that takes the storage as a dependency and follow the same `WithCache` approach that we currently see in the `stored_requests` package.

There’s a lot of logic that’s open for discussion and spread across issues and PRs related to the account service and lazy/eager loading. Perhaps we should first agree on scope and a phased rollout plan so you can meet your timeline before we make some of the larger changes that we have been discussing that will improve testability, readability and extensibility.

It looks like your immediate needs are:

1. Ability to lazy load and refresh file-based stored requests with a LRU freecache-backed caching layer
2. Ability to load and refresh http-based account configs with a LRU freecache-backed caching layer

And the community needs the ability to load account configs from a file (without refresh initially)

Assuming I've captured the requirements correctly, what do you think about this phased rollout plan:
1. (laurb9 #1411) Lazy/eager loading. Needs to be backwards compatible so add config field for loading strategy
2. (laurb9 #1431) Stored request package caching refactor
3. (laurb9 #1426) Stored request package load stored requests/accounts read from file, http or db with ability to use LRU cache
4. (bsardo - in progress) Stored request package fetcher refactor to use ticker tasks
5. (bsardo - in progress) New fetcher and accounts packages with static file(s) fetcher supporting one-time load no refresh (fetcher package introduces storage layer for fetchers, data stored in simple go map with Account values)
6. (shared) Add support to new fetcher and account packages that allows loading accounts from files using LRU cache
7. (laurb9) Add http and db fetcher logic to new fetcher package
8. (shared) Port remaining stored request package caching and fetchers to fetcher package and account data structures to accounts package

How does that sound? Once we agree on that, we can start to make more traction on everything you have open.

laurb9 commented:
Added per-account control of cache TTLs to show how this would work.
Without accounts defined, this works just like before.

GDPR can be enabled/disabled by account in a similar way. I have a commit for that but didn't seem wise to combine everything here.

laurb9 commented:
@bsardo following up on our discussion, I cleaned up the branch, removed the unnecessary parts, rebased and converted the draft to PR open for review. I've also updated the description to summarize the changes better.

I've put the http API in a second commit because while it seemed easy enough to implement, I haven't had a chance to test it and without cache support it is not production ready, and I can leave it out for a later update if needed.

laurb9 commented on config/accounts.go:
```diff
@@ -0,0 +1,23 @@
+	TTL              DefaultTTLs  `mapstructure:"ttl" json:"ttl"`
+	Events           Events       `mapstructure:"events" json:"events"`
+	CCPA             CCPA         `mapstructure:"ccpa" json:"ccpa"`
+	GDPR             TCF2         `mapstructure:"gdpr" json:"gdpr"`
```
I removed the privacy settings in the latest version of the branch.

laurb9 commented on config/config.go:
```diff
@@ -38,6 +38,8 @@ type Configuration struct {
 	StoredRequests  StoredRequests     `mapstructure:"stored_requests"`
 	CategoryMapping StoredRequestsSlim `mapstructure:"category_mapping"`
+	Accounts        StoredRequestsSlim `mapstructure:"accounts"`
+	DefaultAccount  Account            `mapstructure:"default_account"`
```
Long term, I think it is helpful to have the same settings in the same place, so we know which can be 
overridden in account settings.

I've been thinking about this, and it's possible we might need one but probably two.
- this "DefaultAccount" or rather "UnknownAccount" to be returned as-is for invalid (unknown) accounts.
- "AccountDefaults" that you hinted at in another comment, fills in missing parameters for *valid* (known) accounts, so I could define one account as `{}` and the rest would come from it. This is necessary if the jsons are not complete.

laurb9 replied:
I've renamed DefaultAccount to AccountDefaults, to be more explicit that it is meant to be overridden. 
It is also used to provide defaults for unknown accounts depending on the account required settings.

laurb9 commented on config/accounts.go:
```diff
@@ -0,0 +1,20 @@
+	ID               string       `mapstructure:"id" json:"id"`
+	Disabled         bool         `mapstructure:"disabled" json:"disabled"`
+	PriceGranularity string       `mapstructure:"price_granularity" json:"price_granularity"`
+	TTL              DefaultTTLs  `mapstructure:"ttl" json:"ttl"`
```
We could always change the struct later if need be. Reusing `DefaultTTLs` made applying the account override a simple one-line change, so I went with that. You're right about the naming though.

laurb9 commented on config/accounts.go:
```diff
@@ -0,0 +1,20 @@
+	ID               string       `mapstructure:"id" json:"id"`
+	Disabled         bool         `mapstructure:"disabled" json:"disabled"`
+	PriceGranularity string       `mapstructure:"price_granularity" json:"price_granularity"`
+	TTL              DefaultTTLs  `mapstructure:"ttl" json:"ttl"`
```
Right. Will change to `CacheTTL`

laurb9 commented on config/config.go:
```diff
@@ -537,6 +539,12 @@ func New(v *viper.Viper) (*Configuration, error) {
 	}
 
+	// Migrate global settings to default account
+	if c.AccountRequired == false {
```
It is difficult to preserve the original behavior of account_required, as we introduced new cases:
before:
1. pubID missing, account_required=true   -> FAIL
2. pubID missing, account_required=false -> PASS
now, in addition to those:
3. pubID provided, account not found, account_required=true -> I'd like this to FAIL but must be PASS to support the existing behavior (otherwise we'd need to required all the accounts to exist first)
4. pubID provided, account not found, account_required=false -> PASS but use global setting aka default account.

This suggests a single setting might not be sufficient to cover this.

laurb9 replied:
I've resolved to adding a new explicit setting `ValidAccountRequired` to control the more strict requirement,
as managing it through a combination of flags was non-intuitive. 

It also makes migration easier.

The existing `AccountRequired`=`true` behavior is preserved. We accept any non-empty pubID and create an account for it from the `AccountDefaults` config, allowing the host to start populating accounts progressively (valid accounts always work). When the host is ready to switch to more strict validation, the `ValidAccountRequired` is turned on at which point pubIDs that don't have an explicit account will be rejected.

laurb9 commented on endpoints/openrtb2/auction.go:
```diff
@@ -1282,14 +1287,33 @@ func effectivePubID(pub *openrtb.Publisher) string {
-	} else if _, found := cfg.BlacklistedAcctMap[pubID]; found {
-		// Blacklist account now that we have resolved the value
-		err = error(&errortypes.BlacklistedAcct{Message: fmt.Sprintf("Prebid-server has blacklisted Account ID: %s, please reach out to the prebid server host.", pubID)})
+func (deps *endpointDeps) validateAccount(ctx context.Context, pubID string) (*config.Account, error) {
```
I agree. I went with the assumption that the account configs are complete, but it's not a good one. For example, that would force us to write all the cache settings in each. Thinking of account settings as overrides is better. 

I think the best way to support merging with account defaults is to change back to `json.RawMessage`.

laurb9 commented on stored_requests/backends/file_fetcher/fetcher.go:
```diff
@@ -33,6 +34,24 @@ func (fetcher *eagerFetcher) FetchRequests(ctx context.Context, requestIDs []str
 }
 
+// FetchAccount fetches the host account configuration for a publisher
+func (fetcher *eagerFetcher) FetchAccount(ctx context.Context, accountID string) (*config.Account, error) {
```
Agree that it makes more sense to separate the unmarshaling. In this case it helps with applying defaults, as by the time we have the native struct it is too late to do that.

laurb9 commented on config/config.go:
```diff
@@ -38,6 +38,8 @@ type Configuration struct {
 	DataCache       DataCache          `mapstructure:"datacache"`
 	StoredRequests  StoredRequests     `mapstructure:"stored_requests"`
 	CategoryMapping StoredRequestsSlim `mapstructure:"category_mapping"`
+	Accounts        StoredRequestsSlim `mapstructure:"accounts"`
```
Ah... I assumed (I know, I know) it would be covered by `(StoredRequestsSlim*) validate()` but I am starting to think that might not actually exist. 
So this means probably CategoryMapping and VideoRequests aren't validated either.

bsardo commented on config/accounts.go:
```diff
@@ -0,0 +1,20 @@
+	Analytics        PubAnalytics `mapstructure:"analytics" json:"analytics"`
+}
+
+// Events contains account-level controls for win/impression event support
+type Events struct {
+	Enabled bool `mapstructure:"enabled" json:"enabled"`
+}
```
Let's remove this for now. We can add it back in later when we support events.

laurb9 commented:
Updates:
- `AccountFetcher` returns `json.RawMessage` and account defaults are applied in `getAccount` (previously named `validateAccount`)
- Implemented other requested changes.

laurb9 commented on config/accounts.go:
```diff
@@ -0,0 +1,23 @@
+// Account represents a publisher account configuration
+type Account struct {
+	ID               string       `mapstructure:"id" json:"id"`
+	Disabled         bool         `mapstructure:"disabled" json:"disabled"`
```
That was indeed the plan, but now that we can merge with the account defaults, we can do it either way (Enabled or Disabled)

laurb9 commented on endpoints/openrtb2/auction.go:
```diff
@@ -1282,14 +1287,33 @@ func effectivePubID(pub *openrtb.Publisher) string {
+	var account *config.Account
+	var err error
+	if account, err = deps.accounts.FetchAccount(ctx, pubID); err != nil || account == nil {
+		if deps.cfg.DefaultAccount.Disabled || (deps.cfg.AccountRequired && pubID == pbsmetrics.PublisherUnknown) {
```
This avoids breaking a deployment that has an existing `account_required: true` with the original meaning of "just pass something" (number 3 in https://github.com/prebid/prebid-server/pull/1426#discussion_r470926509). In that case, I think the two will be different:

- account_required: true (existing setting)
- default_account.disabled: false (to return default active accounts)
However, we still want to not reject accounts with no id.

laurb9 replied:
Updated, see https://github.com/prebid/prebid-server/pull/1426#discussion_r476972887

mansinahar commented on stored_requests/backends/file_fetcher/fetcher.go:
```diff
@@ -33,6 +33,19 @@ func (fetcher *eagerFetcher) FetchRequests(ctx context.Context, requestIDs []str
 }
 
+// FetchAccount fetches the host account configuration for a publisher
+func (fetcher *eagerFetcher) FetchAccount(ctx context.Context, accountID string) (json.RawMessage, error) {
```
Please consider adding unit test coverage for this

mansinahar commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,6 +81,49 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+// {
+//   "<accountID>": { ... account data ... }
+// }
+func (fetcher *HttpFetcher) FetchAccount(ctx context.Context, accountID string) (json.RawMessage, error) {
```
Please consider adding unit test coverage for this

laurb9 replied:
http_fetcher is not production ready without the ability to cache, so I will remove it instead.

mansinahar commented on config/config.go:
```diff
@@ -578,6 +588,13 @@ func New(v *viper.Viper) (*Configuration, error) {
+	// Update account defaults and generate base json for patch
+	c.AccountDefaults.CacheTTL = c.CacheURL.DefaultTTLs // comment this out to set explicitly in config
+	var err error
+	if c.AccountDefaultsJSON, err = json.Marshal(c.AccountDefaults); err != nil {
```
If we're always gonna construct the `AccountDefaultsJSON` from the `AccountDefaults` struct then former should likely _not_ be a config option. I can only see a reason for it being a config option if we want the host to be able to provide the account defaults as a json string. They can do that right now as `AccountDefaultsJSON` is a config option but we always override it

laurb9 replied:
I can look into alternatives if this is really a deal breaker, but I want to point out that I followed the same implementation done a few lines above for [GDPR.NonStandardPublisherMap](https://github.com/prebid/prebid-server/blob/master/config/config.go#L586-L591), [BlacklistedAcctMap](https://github.com/prebid/prebid-server/blob/master/config/config.go#L600-L605) and [BlacklistedApps](https://github.com/prebid/prebid-server/blob/master/config/config.go#L593-L598) which were accepted about a year ago.

laurb9 replied:
Added commit dc8a74d to make the json a private config member, exposed to other packages via accessor method.

laurb9 commented on config/config.go:
```diff
@@ -113,6 +122,15 @@ func (cfg *Configuration) validate() configErrors {
+	}
+	if cfg.AccountDefaults.Disabled {
+		glog.Warning(`With account_defaults.disabled=true, host-defined accounts must have "disabled":false. All other requests will be rejected.`)
+		if !cfg.ValidAccountRequired {
```
`ValidAccountRequired` was removed yesterday per our discussion, see https://github.com/prebid/prebid-server/pull/1426#discussion_r476972887

laurb9 commented on config/accounts.go:
```diff
@@ -0,0 +1,14 @@
+type Account struct {
+	ID               string       `mapstructure:"id" json:"id"`
+	Disabled         bool         `mapstructure:"disabled" json:"disabled"`
+	PriceGranularity string       `mapstructure:"price_granularity" json:"price_granularity"`
```
The account config structure was trying to match the corresponding Java account config model:[org/prebid/server/settings/model/Account.java](https://github.com/rubicon-project/prebid-server-java/blob/master/src/main/java/org/prebid/server/settings/model/Account.java)

I have removed GDPR settings since then, so I can remove everything else that is not relevant to the functionality added here.

laurb9 commented on endpoints/openrtb2/auction.go:
```diff
@@ -1282,14 +1287,45 @@ func effectivePubID(pub *openrtb.Publisher) string {
+	var err error
+	if accountJSON, err = deps.accounts.FetchAccount(ctx, pubID); err != nil || accountJSON == nil {
+		// pubID does not reference a valid account, or is empty
+		if deps.cfg.AccountRequired &&
```
Will change.

The assumption I made was that requests without a publisher ID should be rare. As it so happens, this made the code more compact as that error is only generated in one place (with the precheck, it would be in two places).

laurb9 commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -346,12 +368,23 @@ func (gr *getResponseFromDirectory) doRequest(t *testing.T, requestData []byte)
+		BlacklistedAccts:   []string{"bad_acct"},
+		BlacklistedAcctMap: map[string]bool{"bad_acct": true},
+		AccountRequired:    gr.accountReq,
+		AccountDefaults:    config.Account{Disabled: gr.validAccountReq},
```
should be accountDefaultsDisabled, thanks

laurb9 commented on config/config.go:
```diff
@@ -592,6 +617,9 @@ func New(v *viper.Viper) (*Configuration, error) {
 	}
 
+	if len(c.BlacklistedAccts) > 0 && c.BlacklistedAccts[0] != "" {
+		glog.Warningf(`Instead of blacklisted_accts, consider using account setting "disabled: true" or global setting "account_required: true".`)
```
I will remove the warning. 
If the host will have defined accounts, it will be obvious they can be disabled individually.
Otherwise, not much point in starting with the blacklisted ones.

laurb9 commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -103,7 +104,7 @@ func TestExplicitUserId(t *testing.T) {
 	// NewMetrics() will create a new go_metrics MetricsEngine, bypassing the need for a crafted configuration set to support it.
 	// As a side effect this gives us some coverage of the go_metrics piece of the metrics engine.
```
The code paths in `getAccount()` are exercised by `TestRejectAccountRequired` as before, when the method was `validateAccount()`. It's what I used to ensure the existing behavior is maintained with the changes.

I can implement another set directed at `getAccount()` but it will be a copy of the cases from that other test, with the similar setup (because it needs endpoint deps).

laurb9 commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,6 +81,49 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+	if len(accountID) == 0 {
+		return nil, nil
+	}
+	httpReq, err := http.NewRequest("GET", fmt.Sprintf(`account=["%s"]`, accountID), nil)
```
endpoint was accidentally missing :)

laurb9 commented on stored_requests/backends/http_fetcher/fetcher.go:
```diff
@@ -81,6 +81,49 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(`Error fetching account "%s" via http: unexpected response status %d`, accountID, httpResp.StatusCode)
+	}
+	var accounts accountsResponseContract
```
To have consistent interface with requests and leave open the ability to query multiple ids.

laurb9 commented on endpoints/openrtb2/auction_test.go:
```diff
@@ -264,6 +265,27 @@ func TestRejectAccountRequired(t *testing.T) {
 			accountReq:    true,
 		},
+		{
+			// Account is required, was provided, not blacklisted, but is not valid (not defined by host)
```
No, but I should reword this. It is the same situation as above, except now we expect strict validation of accounts.

laurb9 commented on stored_requests/multifetcher.go:
```diff
@@ -36,6 +36,17 @@ func (mf MultiFetcher) FetchRequests(ctx context.Context, requestIDs []string, i
+func (mf MultiFetcher) FetchAccount(ctx context.Context, accountID string) (json.RawMessage, error) {
+	for _, f := range mf {
+		if af, ok := f.(AccountFetcher); ok {
+			if account, err := af.FetchAccount(ctx, accountID); err == nil {
```
I will implement a slice of errors like in FetchRequests then.

laurb9 commented:
I have addressed most comments except https://github.com/prebid/prebid-server/pull/1426#discussion_r479553373 which I would like a pass on.

Heads up that I had to modify the response handling code in `amp_auction` to handle error slices: https://github.com/prebid/prebid-server/pull/1426/commits/52a5dd07d38811eef901cf826d50aaac60aaee0b#diff-b28fbda364100254648764deee6101ddL161-L176
In order to pass FetchAccount errors in the response, `FetchAccount()` and `getAccount()` now return error slices.
`auction.go` and `video_auction.go` already had code to handle error slices, but surprisingly `amp_auction` did not.

I've removed the http api that I had in the first PR, since it would not be ready without cache support.

bsardo commented on endpoints/openrtb2/auction.go:
```diff
@@ -1282,14 +1287,55 @@ func effectivePubID(pub *openrtb.Publisher) string {
-	} else if _, found := cfg.BlacklistedAcctMap[pubID]; found {
-		// Blacklist account now that we have resolved the value
-		err = error(&errortypes.BlacklistedAcct{Message: fmt.Sprintf("Prebid-server has blacklisted Account ID: %s, please reach out to the prebid server host.", pubID)})
+func (deps *endpointDeps) getAccount(ctx context.Context, pubID string) (account *config.Account, errs []error) {
```
Can we add unit tests for `getAccount` even though some of it is covered in the E2E test?

laurb9 commented on config/config.go:
```diff
@@ -608,6 +631,20 @@ func New(v *viper.Viper) (*Configuration, error) {
+}
+
+// GetAccountDefaultsJSON returns the precompiled JSON form of account_defaults
+func (cfg *Configuration) GetAccountDefaultsJSON() json.RawMessage {
```
Updated in 952bfc1

bsardo commented on config/stored_requests.go:
```diff
@@ -103,17 +115,19 @@ func resolvedStoredRequestsConfig(cfg *Configuration) {
 	amp.CacheEvents.Endpoint = "/storedrequests/amp"
 	amp.HTTPEvents.Endpoint = sr.HTTPEvents.AmpEndpoint
-
-	// Set data types for each section
-	cfg.StoredRequests.dataType = RequestDataType
-	cfg.StoredRequestsAMP.dataType = AMPRequestDataType
-	cfg.StoredVideo.dataType = VideoDataType
-	cfg.CategoryMapping.dataType = CategoryDataType
```
I think I'd rather see this code here than assign `dataType` inside the `validate`. `validate` probably should just validate, not set cfg values.

bsardo reviewed (approved):
LGTM. I also ran manual E2E tests for the 16 cases in `TestGetAccount` in `auction_test.go` and everything looked good.
